`Undefined symbols for architecture arm64` after nix upgrade

This issue is not encountered on main but does happen when upgrading to a newer [nix revision](https://github.com/roc-lang/roc/blob/6710b371a43639d7a480f03acacd8e5b26267251/flake.nix#L5).

This error only happens when building the tests `$ cargo test --release --no-run`, not when doing `cargo build --release`, building/running roc files, nor when using the repl.
Partial error:
```
[...]
error: could not compile `roc_repl_expect` (lib test) due to previous error
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/nix/store/k8ipha35ighvhbigcxh906fylq1w2c5a-rust-default-1.71.1/lib/rustlib/aarch64-apple-darwin/bin:[...]-dead_strip" "-nodefaultlibs"
  = note: Undefined symbols for architecture arm64:
            "std::terminate()", referenced from:
                llvm::itanium_demangle::Node::printAsOperand(llvm::itanium_demangle::OutputBuffer&, llvm::itanium_demangle::Node::Prec, bool) const in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::PODSmallVector<llvm::itanium_demangle::Node*, 32ul>::push_back(llvm::itanium_demangle::Node* const&) in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::SpecialName::printLeft(llvm::itanium_demangle::OutputBuffer&) const in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::CtorVtableSpecialName::printLeft(llvm::itanium_demangle::OutputBuffer&) const in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::NameType::printLeft(llvm::itanium_demangle::OutputBuffer&) const in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::ModuleName::printLeft(llvm::itanium_demangle::OutputBuffer&) const in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                llvm::itanium_demangle::PODSmallVector<llvm::itanium_demangle::PODSmallVector<llvm::itanium_demangle::Node*, 8ul>*, 4ul>::push_back(llvm::itanium_demangle::PODSmallVector<llvm::itanium_demangle::Node*, 8ul>* const&) in libllvm_sys-a632a77be37a2bc1.rlib(ItaniumDemangle.cpp.o)
                ...
[...]
            "operator new(unsigned long)", referenced from:
                _LLVMContextCreate in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                llvm::Module::setModuleInlineAsm(llvm::StringRef) in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                _LLVMAppendBasicBlockInContext in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                _LLVMCreateBuilderInContext in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                _LLVMCreatePassManager in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                _LLVMCreateFunctionPassManagerForModule in libllvm_sys-a632a77be37a2bc1.rlib(Core.cpp.o)
                std::__1::vector<llvm::GenericValue, std::__1::allocator<llvm::GenericValue>>::reserve(unsigned long) in libllvm_sys-a632a77be37a2bc1.rlib(ExecutionEngineBindings.cpp.o)
                ...
          ld: symbol(s) not found for architecture arm64
          clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: could not compile `test_gen` (test "test_gen") due to previous error
Undefined symbols
```


I was able to bisect this to nixpkgs commit [bcbdb800cf7659d6ff36ac114121a056fe8c9656
](https://github.com/NixOS/nixpkgs/commit/bcbdb800cf7659d6ff36ac114121a056fe8c9656).

This commit updates llvmPackages on Darwin from 11 to 16. I'm not yet sure why this change is resulting in the `Undefined symbols for architecture arm64`. The fact that this is only happening when building tests is probably a valuable clue.

--- a/crates/repl_wasm/build-www.sh
+++ b/crates/repl_wasm/build-www.sh
@@ -9,7 +9,7 @@
 # We use this two-step process because Netlify times out if we try to build the Web REPL there.
 
 # https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/
-set -euxo pipefail
+set -exo pipefail
 
 if ! which wasm-pack
 then
@@ -23,9 +23,22 @@ cd $SCRIPT_RELATIVE_DIR
 mkdir -p build
 rm -rf build/*
 
-# We want a release build, but with debug info (to get stack traces for Wasm backend panics)
-# This configuration is called `--profiling`
-wasm-pack build --profiling --target web -- --features console_error_panic_hook
+# c++abi is not needed for wasm-pack and causes an error, see #6303 for more info
+REMOVE_STR="-C link-arg=-lc++abi"
+
+( # start subshell to limit scope of export RUSTFLAGS
+    # Check if RUSTFLAGS contains the string to be removed
+    if [[ $RUSTFLAGS == *"$REMOVE_STR"* ]]; then
+    # Remove the string
+    RUSTFLAGS=$(echo "$RUSTFLAGS" | sed "s/$REMOVE_STR//g")
+
+    export RUSTFLAGS
+    fi
+
+    # We want a release build, but with debug info (to get stack traces for Wasm backend panics)
+    # This configuration is called `--profiling`
+    wasm-pack build --profiling --target web -- --features console_error_panic_hook
+)
 cp -v pkg/roc_repl_wasm.js build
 
 # To disable optimizations while debugging, do `export REPL_DEBUG=1` before running the script

--- a/flake.nix
+++ b/flake.nix
@@ -2,7 +2,7 @@
   description = "Roc flake";
 
   inputs = {
-    nixpkgs.url = "github:nixos/nixpkgs?rev=676fe5e01b9a41fa14aaa48d87685677664104b1";
+    nixpkgs.url = "github:nixos/nixpkgs?rev=886c9aee6ca9324e127f9c2c4e6f68c2641c8256";
 
     # rust from nixpkgs has some libc problems, this is patched in the rust-overlay
     rust-overlay = {
@@ -134,7 +134,9 @@
           shellHook = ''
             export LLVM_SYS_${llvmMajorMinorStr}_PREFIX="${llvmPkgs.llvm.dev}"
             ${aliases}
-          '';
+          '' + pkgs.lib.optionalString (system == "aarch64-darwin") ''
+            export RUSTFLAGS="-C link-arg=-lc++abi"
+          ''; # lc++abi as workaround for github.com/NixOS/nixpkgs/issues/166205, see also github.com/roc-lang/roc/issues/6303
         };
 
         formatter = pkgs.nixpkgs-fmt;

